File: bubble-sort/python/bubble_sort.py
from typing import Any


def swap_list_elements(target_list: list[Any], index_one: int, index_two: int):
    try:
        old_one = target_list[index_one]

        target_list[index_one] = target_list[index_two]
        target_list[index_two] = old_one
    except IndexError:
        print('List index out of range.')


def bubble_sort(unsorted_numbers: list[int]) -> list[int]:
    while not is_sorted(unsorted_numbers):
        for index in range(len(unsorted_numbers) - 1):
            current_number = unsorted_numbers[index]
            next_number = unsorted_numbers[index + 1]

            if current_number > next_number:
                swap_list_elements(unsorted_numbers, index, index + 1)

    return unsorted_numbers


def is_sorted(maybe_sorted_numbers: list[int]) -> bool:
    _sorted = True

    for index in range(len(maybe_sorted_numbers) - 1):
        current_number = maybe_sorted_numbers[index]
        next_number = maybe_sorted_numbers[index + 1]

        if current_number > next_number:
            _sorted = False
            break

    return _sorted


unsorted_list = [1, 19, 22, 1, 2, 5, 3, 99, 127, 12, 4]

sorted_list = bubble_sort(unsorted_list)

print(sorted_list)

File: bubble-sort/rust/bubble_sort.rs
fn main() {
    let ages: Vec<usize> = vec![17, 4, 2, 23, 42, 12, 43, 20, 120, 1209, 12, 3, 5, 1, 5, 11, 1, 2];

    let sorted_ages = bubble_sort(ages);

    println!("{:?}", sorted_ages);
}

fn bubble_sort(mut unsorted_numbers: Vec<usize>) -> Vec<usize> {
    while is_sorted(&unsorted_numbers) == false {
        for idx in 0..unsorted_numbers.len() - 1 {
            let maybe_current_number = unsorted_numbers.get(idx);
            let maybe_next_number = unsorted_numbers.get(idx + 1);

            match (maybe_current_number, maybe_next_number) {
                (Some(current_number), Some(next_number)) => {
                    if current_number > next_number {
                        unsorted_numbers.swap(idx, idx + 1);
                    }
                }

                _ => {}
            }
        }
    }

    unsorted_numbers
}

fn is_sorted(maybe_sorted_numbers: &Vec<usize>) -> bool {
    let mut sorted = true;

    for (idx, current_number) in maybe_sorted_numbers.iter().enumerate() {
        let maybe_next_number = maybe_sorted_numbers.get(idx + 1);

        match maybe_next_number {
            Some(next_number) => {
                if next_number < current_number {
                    sorted = false;
                    break;
                }
            }

            None => ()
        }
    }

    sorted
}

